iree/compiler/Dialect/HAL/IR/HALOpFolders.cpp

```diff
@@ -32,6 +32,16 @@ OpFoldResult TensorCastOp::fold(ArrayRef<Attribute> operands) {
   if (source().getType() == target().getType()) {
     return source();
   }
+
+  // Cast of a cast can use the defining op's source.
+  // This can apply recursively and may bottom out at source == target type.
+  if (auto castOp = source().getDefiningOp<TensorCastOp>()) {
+    auto mutableSource = sourceMutable();
```

Comment:
you don't want to mutate the op in a folder as the folder can run speculatively and you only want to return the folded result (the folder says "this op folds into this value/attribute") - something mutating the op in place should be a canonicalizer

here you'd be checking to see if the source of the defining cast is the same as the target and then just return that castOp.source() (saying "this op folds into the source of a matched cast")

Reply:
Oh, I was referencing https://mlir.llvm.org/docs/Canonicalization/#canonicalizing-with-the-fold-method

> fold has the restriction that no new operations may be created, and only the root operation may be replaced. It allows for updating an operation in-place, or returning a set of pre-existing values (or attributes) to replace the operation with.

> ///  2. They can mutate the operation in place, without changing anything else
///     in the IR. In this case, return the operation itself.

Reply:
that's so sketchy :P I've not seen that before